### PR TITLE
Move ElasticJobListener to JobConfiguration

### LINKED ISSUE #1451: Move ElasticJobListener to JobConfiguration
ElasticJobListener is a bit heavy as a schedule parameter., it's better to configure in JobConfiguration like ErrorHandler.
So JobListener should listener shard item or job level event?

### PR DESCRIPTION
**Pull request by Technoboy-**
Fixes #1451.

### CODE REVIEW

File: elasticjob-api/src/main/java/org/apache/shardingsphere/elasticjob/api/JobConfiguration.java [hunk 1]
```diff
@@ -106,6 +111,8 @@ public static Builder newBuilder(final String jobName, final int shardingTotalCo
         private String jobExecutorServiceHandlerType;
         
         private String jobErrorHandlerType;
+    
+        private final Collection<String> jobListenerTypes = new ArrayList<>();
```
Comment (TeslaCN) [inline elasticjob-api/src/main/java/org/apache/shardingsphere/elasticjob/api/JobConfiguration.java:115]: Collection is unordered. Consider HashSet or other?

File: elasticjob-lite/elasticjob-lite-core/src/main/java/org/apache/shardingsphere/elasticjob/lite/internal/schedule/JobScheduler.java [hunk 2]
```diff
@@ -71,46 +68,44 @@
...
-    public JobScheduler(final CoordinatorRegistryCenter regCenter, final ElasticJob elasticJob, final JobConfiguration jobConfig, final TracingConfiguration<?> tracingConfig,
-                        final ElasticJobListener... elasticJobListeners) {
+    public JobScheduler(final CoordinatorRegistryCenter regCenter, final ElasticJob elasticJob, final JobConfiguration jobConfig, final TracingConfiguration<?> tracingConfig) {
         this.regCenter = regCenter;
-        this.elasticJob = elasticJob;
         elasticJobType = null;
-        this.elasticJobListeners = Arrays.asList(elasticJobListeners);
-        setUpFacade = new SetUpFacade(regCenter, jobConfig.getJobName(), this.elasticJobListeners);
+        final Collection<ElasticJobListener> elasticJobListeners = jobConfig.getJobListenerTypes().stream()
+                .map(type -> ElasticJobListenerFactory.getListener(type)).collect(Collectors.toList());
```
Comment (TeslaCN) [inline elasticjob-lite/elasticjob-lite-core/src/main/java/org/apache/shardingsphere/elasticjob/lite/internal/schedule/JobScheduler.java]: I think replace lambda with method reference 'ElasticJobListenerFactory::getListener' is more elegant.

File: elasticjob-lite/elasticjob-lite-core/src/main/java/org/apache/shardingsphere/elasticjob/lite/internal/schedule/JobScheduler.java [hunk 3]
```diff
@@ -71,46 +68,44 @@
...
-    public JobScheduler(final CoordinatorRegistryCenter regCenter, final String elasticJobType, final JobConfiguration jobConfig, final TracingConfiguration<?> tracingConfig,
-                        final ElasticJobListener... elasticJobListeners) {
+    public JobScheduler(final CoordinatorRegistryCenter regCenter, final String elasticJobType, final JobConfiguration jobConfig, final TracingConfiguration<?> tracingConfig) {
         this.regCenter = regCenter;
-        elasticJob = null;
         this.elasticJobType = elasticJobType;
-        this.elasticJobListeners = Arrays.asList(elasticJobListeners);
-        setUpFacade = new SetUpFacade(regCenter, jobConfig.getJobName(), this.elasticJobListeners);
+        final Collection<ElasticJobListener> elasticJobListeners = jobConfig.getJobListenerTypes().stream()
+                .map(type -> ElasticJobListenerFactory.getListener(type)).collect(Collectors.toList());
```
Comment (TeslaCN) [inline elasticjob-lite/elasticjob-lite-core/src/main/java/org/apache/shardingsphere/elasticjob/lite/internal/schedule/JobScheduler.java]: I think replace lambda with method reference 'ElasticJobListenerFactory::getListener' is more elegant.

File: elasticjob-lite/elasticjob-lite-core/src/test/resources/META-INF/services/org.apache.shardingsphere.elasticjob.infra.listener.ElasticJobListener [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+org.apache.shardingsphere.elasticjob.lite.api.listener.fixture.TestDistributeOnceElasticJobListener
+org.apache.shardingsphere.elasticjob.lite.api.listener.fixture.TestElasticJobListener
+org.apache.shardingsphere.elasticjob.lite.integrate.listener.TestDistributeOnceElasticJobListener
+org.apache.shardingsphere.elasticjob.lite.integrate.listener.TestElasticJobListener
```
Comment (TeslaCN) [inline elasticjob-lite/elasticjob-lite-core/src/test/resources/META-INF/services/org.apache.shardingsphere.elasticjob.infra.listener.ElasticJobListener]: Need a blank line.

**Comment by TeslaCN** [comment]
I think using List is better than using Collection if listeners are in order. Using Collection may be misleading.

File: elasticjob-api/src/main/java/org/apache/shardingsphere/elasticjob/api/JobConfiguration.java (hunk 1 above)
Comment (terrymanu) [reply elasticjob-api/src/main/java/org/apache/shardingsphere/elasticjob/api/JobConfiguration.java:115]: Collection is better, HashSet is not ordered too.

**Review by terrymanu — APPROVED** [review]

**Pull request merged**